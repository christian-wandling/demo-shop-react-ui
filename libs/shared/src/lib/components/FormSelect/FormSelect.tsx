import { Controller, useFormContext } from 'react-hook-form';
import { FormSelectProps } from '../../models/formSelectProps';

export const FormSelect = ({
  name,
  label,
  options,
  required = false,
  disabled = false,
  className,
  'data-testid': dataTestId,
  defaultValue = '',
}: FormSelectProps) => {
  const {
    control,
    formState: { errors },
  } = useFormContext();
  const errorMessage = errors[name]?.message as string | undefined;

  return (
    <div className={className}>
      <label
        htmlFor={name}
        className={`block text-sm font-medium leading-6 ${disabled ? 'text-gray-500' : 'text-gray-900'}`}>
        {label} {required && <span className="required">*</span>}
      </label>
      <Controller
        name={name}
        control={control}
        defaultValue={defaultValue}
        render={({ field }) => (
          <select
            id={name}
            disabled={disabled}
            className={`block w-full rounded-md border-0 py-1.5 shadow-sm ring-1 ring-inset ring-gray-300 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:max-w-xs sm:text-sm sm:leading-6 ${
              disabled ? 'text-gray-500 bg-gray-200' : 'text-gray-900'
            } ${errorMessage ? 'error' : ''} ${className || ''}`}
            data-testid={dataTestId}
            {...field}>
            <option disabled selected value="">
              Select {label}
            </option>
            {options.map(option => (
              <option key={option.value} value={option.value}>
                {option.label}
              </option>
            ))}
          </select>
        )}
      />
      {errorMessage && <div className="error-message">{errorMessage}</div>}
    </div>
  );
};
