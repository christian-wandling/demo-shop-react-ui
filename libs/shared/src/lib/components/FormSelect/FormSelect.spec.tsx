import { beforeEach, describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { FormSelect } from './FormSelect';
import { useFormContext } from 'react-hook-form';

vi.mock('react-hook-form', () => ({
  useFormContext: vi.fn(),
  Controller: vi.fn(({ render }) => render({ field: { onChange: vi.fn(), value: '' } })),
}));

describe('FormSelect', () => {
  const mockControl = { control: 'mock-control' };
  const mockErrors = {};

  const defaultProps = {
    name: 'testSelect',
    label: 'Test Select',
    options: [
      { value: 'option1', label: 'Option 1' },
      { value: 'option2', label: 'Option 2' },
    ],
  };

  beforeEach(() => {
    vi.clearAllMocks();
    (useFormContext as any).mockReturnValue({
      control: mockControl,
      formState: { errors: mockErrors },
    });
  });

  it('renders with basic props', () => {
    render(<FormSelect {...defaultProps} />);

    expect(screen.getByLabelText(/Test Select/)).toBeTruthy();
    expect(screen.getByText('Select Test Select')).toBeTruthy();
    expect(screen.getByText('Option 1')).toBeTruthy();
    expect(screen.getByText('Option 2')).toBeTruthy();
  });

  it('matches the snapshot', () => {
    const { baseElement } = render(<FormSelect {...defaultProps} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('shows required asterisk when required is true', () => {
    render(<FormSelect {...defaultProps} required={true} />);

    const label = screen.getByText('*');
    expect(label).toBeTruthy();
  });

  it('applies disabled styling when disabled is true', () => {
    render(<FormSelect {...defaultProps} disabled={true} />);

    const select: HTMLSelectElement = screen.getByRole('combobox');
    expect(select.disabled).toBe(true);

    const label = screen.getByText('Test Select');
    expect(label.className).toContain('text-gray-500');
  });

  it('applies additional className when provided', () => {
    render(<FormSelect {...defaultProps} className="custom-class" />);

    const selectContainer = screen.getByRole('combobox').parentElement;
    expect(selectContainer?.classList).toContain('custom-class');
  });

  it('displays error message when there is an error', () => {
    const errorMessage = 'This field is required';
    (useFormContext as any).mockReturnValue({
      control: mockControl,
      formState: { errors: { testSelect: { message: errorMessage } } },
    });

    render(<FormSelect {...defaultProps} />);

    expect(screen.getByText(errorMessage)).toBeTruthy();
    const select = screen.getByRole('combobox');
    expect(select.className).toContain('error');
  });

  it('renders all provided options', () => {
    const manyOptions = [
      { value: 'option1', label: 'Option 1' },
      { value: 'option2', label: 'Option 2' },
      { value: 'option3', label: 'Option 3' },
      { value: 'option4', label: 'Option 4' },
    ];

    render(<FormSelect {...defaultProps} options={manyOptions} />);

    manyOptions.forEach(option => {
      expect(screen.getByText(option.label)).toBeTruthy();
    });
  });
});
