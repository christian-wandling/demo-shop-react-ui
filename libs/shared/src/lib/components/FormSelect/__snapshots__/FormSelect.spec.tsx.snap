// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormSelect > matches the snapshot 1`] = `
<body>
  <div>
    <div>
      <label
        class="block text-sm font-medium leading-6 text-gray-900"
        for="testSelect"
      >
        Test Select
         
      </label>
      <select
        class="
              block w-full rounded-md border-0 py-1.5 shadow-sm ring-1 ring-inset ring-gray-300 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:max-w-xs sm:text-sm sm:leading-6
              text-gray-900
               
            "
        id="testSelect"
      >
        <option
          value=""
        >
          Select 
          Test Select
        </option>
        <option
          value="option1"
        >
          Option 1
        </option>
        <option
          value="option2"
        >
          Option 2
        </option>
      </select>
    </div>
  </div>
</body>
`;
